pkg/kubelet/kubelet.go

```diff
@@ -2972,22 +3070,47 @@ func (kl *Kubelet) CheckpointContainer(
 		return fmt.Errorf("container %v not found", containerName)
 	}
 
-	options.Location = filepath.Join(
+	checkpointDirectory := filepath.Join(
 		kl.getCheckpointsDir(),
+		string(podUID),
+	)
+
+	checkPointDirectoryInfo, err := os.Lstat(checkpointDirectory)
+	if err == nil && !checkPointDirectoryInfo.Mode().IsDir() {
+		return fmt.Errorf("checkpoint destination %s needs to be a directory", checkpointDirectory)
+	} else if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error stating checkpoint destination %s:%w", checkpointDirectory, err)
+	} else if err != nil && os.IsNotExist(err) {
+		if err = os.Mkdir(checkpointDirectory, 0700); err != nil {
+			return fmt.Errorf("error creating checkpoint directory %s:%w", checkpointDirectory, err)
+		}
+	}
+
+	kl.maxCheckpointsPerContainerLock.Lock()
+	options.Location = filepath.Join(
+		checkpointDirectory,
 		fmt.Sprintf(
-			"checkpoint-%s-%s-%s.tar",
+			"checkpoint-%s-%s-%d-%0*d.tar",
 			podFullName,
 			containerName,
-			time.Now().Format(time.RFC3339),
+			time.Now().UnixNano(),
```

Comment:
I thought I submitted this comment two days ago. Trying once more:


This PR has gone though multiple iterations. The first iterations was using the existing file name with the human readable and the number of checkpoints was tracked in a JSON file. That was then changed to work without a JSON file using `stat()`. The time resolution using `stat()` on all file systems was questioned so the goal was to use a regex and file system sorting. The problem with the human readable file name was that it is really hard to capture it via regex if timezones are used in the file name. In combination with `pod-name` `container-name` collisions during regex or file system level sorting the current implementation using an integer time stamp and a counter.

As the second resolution is to coarse to be unique (without the counter) I switched to something else. I thought about micro or nano seconds, and settled for nano seconds because it will be just as unreadable as micro seconds.

We have written the tool `checkpointctl` which will display all information, including time created and checkpointed, so if wanted it can be displayed with the help of an additional tool.

The current approach gives us file system based sorting by using unique names in combination with podUID, nano seconds and node wide counter in the file name. No external JSON file is any more needed or no need to ready the result of `stat()` which might be to coarse, depending on the used file system.

With the old implementation if would have been possible to create two checkpoints of different pod/containers at the same time and once would replace another. The current implementation does not have this problem and easier to sort.

Reply:
looking up... the dreaded windows aversion to colon due to their drive letter `:` format..

sounds like you had a lot of reasons..